put_item.go

```diff
@@ -10,12 +10,37 @@ import (
 	"github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
 )
 
+type PutOption func(*dynamodb.PutItemInput) error
+
+// Enables concurrency control by using an optimistic lock
+// https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBMapper.OptimisticLocking.html
+//
+// Provide key field acts as a version number (Usually called Version)
+// GetItem retrieves current version number and you can update the item if the version number in DynamoDB hasn't changed
+// Each update increments the version number and if the update fails fetch the record again to get latest version number and try again
+func WithOptimisticLock(key string, currentVersion uint) PutOption {
```

Comment:
Do we need to specify key as DynamoDBVersionAttribute somewhere?

And from the doc, it's auto updating the key value, do we need to call increment + 1 manually like this?

`try {
    dynamoDBMapper.save(order);
} catch (OptimisticLockingException e) {
    System.out.println("Conflict detected, retrying...");
    order = dynamoDBMapper.load(Order.class, order.getOrderId()); // Fetch latest
    order.setQuantity(order.getQuantity() + 1);
    dynamoDBMapper.save(order);
}`

Reply:
To put it simply, Go SDK is more low level, we need to implement the behavior ourself using ConditionExpression, which is exactly what the JavaSDK abstraction does.

Reply:
Okay